src/app/shared/issue/title-editor/title-editor.component.ts

```diff
@@ -0,0 +1,134 @@
+
+  ngOnInit() {
+    if (this.initialTitle !== undefined) {
+      this.titleField.setValue(this.initialTitle);
+    }
+
+    if (this.titleField === undefined || this.titleForm === undefined || this.id === undefined) {
+      throw new Error("Title Editor's compulsory properties are not defined.");
+    }
+
+    this.titleField.setValidators([Validators.required, Validators.maxLength(this.maxLength)]);
+    this.history = new UndoRedo<textEntry>(
+      75,
+      () => {
+        return {
+          text: this.titleInput.nativeElement.value,
+          selectStart: this.titleInput.nativeElement.selectionStart,
+          selectEnd: this.titleInput.nativeElement.selectionEnd
+        };
+      },
+      500
+    );
+  }
+
+  onKeyPress(event: KeyboardEvent) {
+    if (this.isUndo(event)) {
+      event.preventDefault();
+      this.undo();
+      return;
+    } else if (this.isRedo(event)) {
+      this.redo();
+      event.preventDefault();
+      return;
+    }
+  }
+
+  handleBeforeInputChange(event: InputEvent): void {
+    switch (event.inputType) {
+      case 'historyUndo':
+      case 'historyRedo':
+        // ignore these events that doesn't modify the text
+        event.preventDefault();
+        break;
+      case 'insertFromPaste':
+        this.history.forceSave(null, true, false);
+        break;
+
+      default:
+        this.history.updateBeforeChange();
+    }
+  }
+
+  handleInputChange(event: InputEvent): void {
+    switch (event.inputType) {
+      case 'historyUndo':
+      case 'historyRedo':
+        // ignore these events that doesn't modify the text
+        event.preventDefault();
+        break;
+      case 'insertFromPaste':
+        // paste events will be handled exclusively by handleBeforeInputChange
+        break;
+
+      default:
+        this.history.createDelayedSave();
+    }
+  }
+
+  private undo(): void {
+    const entry = this.history.undo();
+    if (entry === null) {
+      return;
+    }
+    this.titleField.setValue(entry.text);
+    this.titleInput.nativeElement.setSelectionRange(entry.selectStart, entry.selectEnd);
+  }
+
+  private redo(): void {
+    const entry = this.history.redo();
+    if (entry === null) {
+      return;
+    }
+    this.titleInput.nativeElement.value = entry.text;
+    this.titleInput.nativeElement.setSelectionRange(entry.selectStart, entry.selectEnd);
+  }
+
+  private isUndo(event: KeyboardEvent) {
+    // prevents undo from firing when ctrl shift z is pressed
+    if (navigator.platform.indexOf('Mac') === 0) {
+      return event.metaKey && event.code === 'KeyZ' && !event.shiftKey;
+    }
+    return event.ctrlKey && event.code === 'KeyZ' && !event.shiftKey;
+  }
+
+  private isRedo(event: KeyboardEvent) {
+    if (navigator.platform.indexOf('Mac') === 0) {
+      return event.metaKey && event.shiftKey && event.code === 'KeyZ';
+    }
+    return (event.ctrlKey && event.shiftKey && event.code === 'KeyZ') || (event.ctrlKey && event.code === 'KeyY');
+  }
+}
```

Comment:
I think this code is repeated from `comment-editor.component.ts`. Perhaps these functions can be in `undoredo.model.ts`?

Reply:
Good suggestion, I think `isUndo` and `isRedo` can definitely be moved into `undoredo.model.ts`, perhaps as a static method. However, I'm not sure if the same should be done for the `undo` and `redo` methods, as I personally feel that the handling of values from `UndoRedo::undo`/`UndoRedo::redo` should be done in the components using the `UndoRedo` model, in the interests of keeping `UndoRedo` more general. Could I perhaps ask for your thoughts on this?

Reply:
Leaving the undo and redo methods as it is is okay too